Module-2 Homework-2 API/DTO/Internal models/Mappers

wursta opened the pull request.

wursta committed:
Module-2 Homework-3 API

Добавлена спецификация API в виде yaml файла на основе OpenAPI.

Добавлен модуль fm-api-v1 на основе multiplatform с генератором транспортных моделей по openapi спецификации.

Написана обработка поля-дескриминтора requestType для kotlinx-serialization.

Написаны тесты на сериализацию/десериализацию транспортных моделей.

wursta committed:
Module-2 Homework-3 API

Использован отрефакторенный код создания сериализатора с практики.
Доработана спецификация для сущности трат (Expense)
Написаны тесты на DTO для запросов-ответов по картам и тратам.

wursta committed:
Module-2 Homework-3 Внутренние модели и мапперы

Создан модуль fm-common для внутренних моделей.
Создан модуль fm-mappers для мапперов из DTO во внутренние модели и обратно.
Создан маппер для транспортных моделей сущности "Банковская карта" и написаны unit-тесты на него.

wursta committed:
Module-2 Homework-3 Внутренние модели и мапперы

Переименование модуля fm-api-v1 в fm-api.
Доработка внутренних моделей, мапперов и api.

evgnep commented on specs/v1/schemas/Bank.yaml:
```diff
@@ -0,0 +1,16 @@
+BankGUID:
```
Эти файлы можно оформлять как опенапи - указывать заголовок и т.п.

```
openapi: 3.0.4
info:
  title: "Personal Financial Management Service"
  version: 1.0.0
paths: {}
components:
  schemas:
     ....
```

Тогда будет работать автокомплит в идее и просмотр структуры как опенапи

evgnep commented on specs/v1/schemas/Bank.yaml:
```diff
@@ -0,0 +1,16 @@
+  description: Объект описывает свойства банка
+  properties:
+    guid:
+      $ref: './Bank.yaml#/BankGUID'
```
по идее ссылки внутри файла можно делать без указания самого файла, просто относительная ссылка

evgnep commented on fm-common/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/build/
```
а зачем местный гитигнор? По идее хватает одного на верхнем уровне

evgnep commented on fm-api/src/commonMain/kotlin/local/learning/api/serialization/utils/jsonSerializer.kt:
```diff
@@ -0,0 +1,45 @@
+import local.learning.api.models.*
+
+internal val infos = listOf(
+    // Cards requests
```
я смотрю, зашло вам то, что мы на практике делали :)

evgnep commented on fm-common/src/commonMain/kotlin/models/expense/Expense.kt:
```diff
@@ -0,0 +1,14 @@
+data class Expense(
+    var guid: ExpenseGuid = ExpenseGuid.NONE,
+    var createDT: Instant = INSTANT_NONE,
+    var amount: Float = 0F,
```
не советую использовать float для денег. Используйте BigDecimal лучше, меньше шансов напороться на проблемы

evgnep commented on fm-common/src/commonMain/kotlin/utils.kt:
```diff
@@ -0,0 +1,7 @@
+
+import kotlinx.datetime.Instant
+
+internal val INSTANT_NONE = Instant.fromEpochMilliseconds(Long.MIN_VALUE)
```
а точно internal? Ведь тогда эти константы недоступны вне модуля, а в моделях логики по идее нет...

evgnep commented on fm-mappers/build.gradle.kts:
```diff
@@ -0,0 +1,33 @@
+            }
+        }
+        @Suppress("UNUSED_VARIABLE")
+        val jvmTest by getting {
```
а точно это надо? У нас же нет тестов jvm

evgnep commented on fm-mappers/src/commonMain/kotlin/CardMappersFromTransport.kt:
```diff
@@ -0,0 +1,98 @@
+        owner = this.owner ?: "",
+        bankGuid = this.bank.toBankGuid()
+    ).also {
+        if (!it.isNumberValid()) {
```
вообще у нас по плану валидация идет в рамках бизнес-цепочки через один модуль... Так что возможно это туда переедет

evgnep commented on specs/v1/openapi.yaml:
```diff
@@ -0,0 +1,404 @@
+openapi: 3.0.4
```
Раз уж взялись бить на отдельные файлы, этот файл все равно слишком большой и я б его тоже разбил. К примеру можно вынести разные CRUD в разные файлы. Или вообще каждый метод в отдельный файл

wursta committed:
Module-2 Homework-3 Доработки после код ревью

Замена типа полей в которых передаются "деньги" с float на double в моделях трат в openapi.
Во внутренних моделях для "денег" использован тип BigDecimal.
